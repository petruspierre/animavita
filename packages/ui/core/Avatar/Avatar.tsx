import React, {useRef} from 'react';
import {Pressable, ImageProps} from 'react-native';
import styled from 'styled-components/native';
import Menu, {MenuItem, MenuDivider} from 'react-native-material-menu';
import {px2ddp} from '@animavita/theme';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useNavigation} from '@react-navigation/native';

const AVATAR_SIZE = 18;

const StyledAvatar = styled.Image<Pick<AvatarProps, 'width' | 'height'>>`
  border-radius: ${px2ddp(AVATAR_SIZE) / 2}px;
  width: ${({width}) => width || px2ddp(AVATAR_SIZE)}px;
  height: ${({height}) => height || px2ddp(AVATAR_SIZE)}px;
`;

const StyledMenu = styled(Menu)`
  margin-top: ${({width}) => width || px2ddp(AVATAR_SIZE)}px;
`;

export interface AvatarProps extends ImageProps {
  width?: number;
  height?: number;
  menu?: boolean;
}

interface MenuRef {
  hide: () => void;
  show: () => void;
}

const Avatar: React.FC<AvatarProps> = props => {
  const menu = useRef<MenuRef>(null);
  const navigation = useNavigation();

  const hideMenu = () => {
    if (menu.current) {
      menu.current.hide();
    }
  };

  const showMenu = () => {
    if (props.menu && menu.current) {
      menu.current.show();
    }
  };

  const handleLogout = async () => {
    await AsyncStorage.clear();
    navigation.navigate('SignUp');
  };

  return (
    <StyledMenu
      ref={menu}
      button={
        <Pressable onPress={showMenu}>
          <StyledAvatar {...props} testID="avatar" />
        </Pressable>
      }>
      <MenuItem onPress={hideMenu} disabled>
        Configurações
      </MenuItem>
      <MenuDivider />
      <MenuItem onPress={handleLogout}>Sair</MenuItem>
    </StyledMenu>
  );
};

export default Avatar;
