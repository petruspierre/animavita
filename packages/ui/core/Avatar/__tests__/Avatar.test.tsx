import React from 'react';
import {fireEvent, act} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import AsyncStorage from '@react-native-async-storage/async-storage';

import {Mount, ThemeContext} from '../../../tests/helpers';
import Avatar from '../Avatar';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockedNavigate,
  }),
}));

const defaultProps = {
  source: {uri: 'imageurl.com'},
};

const mountFactory = propOverrides => {
  const avatarProps = {...defaultProps, ...propOverrides};

  const element = Mount(<Avatar {...avatarProps} />);

  return element;
};

describe('Avatar', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('custom dimensions', () => {
    it('renders with custom width', () => {
      const {getByTestId} = mountFactory({width: 20});

      expect(getByTestId('avatar')).toHaveStyle({width: 20});
    });

    it('renders with custom height', () => {
      const {getByTestId} = mountFactory({height: 20});

      expect(getByTestId('avatar')).toHaveStyle({height: 20});
    });
  });

  describe('menu enabled', () => {
    it('logout application', async () => {
      const {getByTestId, getByText} = mountFactory({menu: true});

      await act(async () => {
        await fireEvent.press(getByTestId('avatar'));
        await fireEvent.press(getByText('Sair'));
      });

      expect(AsyncStorage.clear).toHaveBeenCalled();
      expect(AsyncStorage.clear).toHaveBeenCalledTimes(1);
      expect(mockedNavigate).toHaveBeenCalled();
      expect(mockedNavigate).toHaveBeenCalledTimes(1);
    });
  });

  it('renders correctly', () => {
    const tree = renderer.create(ThemeContext(<Avatar {...defaultProps} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
