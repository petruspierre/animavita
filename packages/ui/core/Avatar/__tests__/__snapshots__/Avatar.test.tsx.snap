// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "uri": "imageurl.com",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
          ]
        }
        testID="avatar"
      />
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "marginTop": 36,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "height": 48,
                    "justifyContent": "center",
                    "maxWidth": 248,
                    "minWidth": 124,
                  },
                  undefined,
                ]
              }
            >
              <Text
                ellipsizeMode="clip"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "400",
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#bdbdbd",
                    },
                    undefined,
                  ]
                }
              >
                Configurações
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                },
                Object {
                  "borderBottomColor": "rgba(0,0,0,0.12)",
                },
              ]
            }
          />
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "height": 48,
                    "justifyContent": "center",
                    "maxWidth": 248,
                    "minWidth": 124,
                  },
                  undefined,
                ]
              }
            >
              <Text
                ellipsizeMode="clip"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "400",
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                    },
                    false,
                    undefined,
                  ]
                }
              >
                Sair
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
